coder-sdk/client.go

```diff
@@ -1,46 +1,265 @@
+	ImageTags(ctx context.Context, imageID string) ([]ImageTag, error)
+
+	// ImageTagByID fetch an image tag by ID.
+	ImageTagByID(ctx context.Context, imageID, tagID string) (*ImageTag, error)
+
+	// CreateAPIToken creates a new APIToken for making authenticated requests to Coder Enterprise.
+	CreateAPIToken(ctx context.Context, userID string, req CreateAPITokenReq) (string, error)
+
+	// APITokens fetches all APITokens owned by the given user.
+	APITokens(ctx context.Context, userID string) ([]APIToken, error)
+
+	// APITokenByID fetches the metadata for a given APIToken.
+	APITokenByID(ctx context.Context, userID, tokenID string) (*APIToken, error)
+
+	// DeleteAPIToken deletes an APIToken.
+	DeleteAPIToken(ctx context.Context, userID, tokenID string) error
+
+	// RegenerateAPIToken regenerates the given APIToken and returns the new value.
+	RegenerateAPIToken(ctx context.Context, userID, tokenID string) (string, error)
+
+	// APIVersion parses the coder-version http header from an authenticated request.
+	APIVersion(ctx context.Context) (string, error)
+
+	// WorkspaceProviderByID fetches a workspace provider entity by its unique ID.
+	WorkspaceProviderByID(ctx context.Context, id string) (*WorkspaceProvider, error)
+
+	// WorkspaceProviders fetches all workspace providers known to the Coder control plane.
+	WorkspaceProviders(ctx context.Context) ([]WorkspaceProvider, error)
+
+	// CreateWorkspaceProvider creates a new WorkspaceProvider entity.
+	CreateWorkspaceProvider(ctx context.Context, req CreateWorkspaceProviderReq) (*CreateWorkspaceProviderRes, error)
+
+	// DeleteWorkspaceProviderByID deletes a workspace provider entity from the Coder control plane.
+	DeleteWorkspaceProviderByID(ctx context.Context, id string) error
+}
+
+// ClientOptions contains options for the Coder SDK Client.
+type ClientOptions struct {
+	// BaseURL is the root URL of the Coder installation.
 	BaseURL *url.URL
-	Token   string
+
+	// Client is the http.Client to use for requests (optional).
+	// If omitted, the http.DefaultClient will be used.
+	HTTPClient *http.Client
+
+	// Token is the API Token used to authenticate
+	Token string
 }
 
-// newHTTPClient creates a default underlying http client and sets the auth cookie.
-//
-// NOTE:
-// As we do not specify a custom transport, the default one from the stdlib will be used,
-// resulting in a persistent connection pool.
-// We do not set a timeout here as it could cause issue with the websocket.
-// The caller is expected to set it when needed.
-//
-// WARNING:
-// If the caller sets a custom transport to set TLS settings or a custom CA the default
-// pool will not be used and it might result in a new dns lookup/tls session/socket begin
-// established each time.
-func (c Client) newHTTPClient() (*http.Client, error) {
-	jar, err := cookiejar.New(nil)
-	if err != nil {
-		return nil, err
+// NewClient creates a new default Coder SDK client.
+func NewClient(opts ClientOptions) (c Client, err error) {
```

Comment:
@cmoog What's your opinion here? I know a common practice is to [return the struct rather than an interface](https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8), which would make this change less invasive. My goal here was to force the use of a constructor, so that we can better control mutable vs immutable state, and so that we can handle alternative construction methods (e.g. create a Client with either an API Token or a username/password combination).

If we return the struct, then we should make it a public type so that it can be assigned (e.g. `var x coder.DefaultClient = abc`), but this means users have to make sure to do that instead of creating an empty Client (e.g. `client := &coder.DefaultClient{}`) which would be useless as it'd be missing the necessary baseURL etc settings

Reply:
I'm not too opinionated here, so I'd generally want us to optimize for the simplest, lowest maintenance path. As far as matching other libraries, as you pointed out with Kubernetes (and [AWS](https://github.com/aws/aws-sdk-go/blob/master/aws/client/client.go) and [GCP](https://github.com/googleapis/google-cloud-go/blob/master/spanner/client.go#L78:6))  it is very standard for packages to expose a public struct type for accessing an underlying HTTP API.